Add doc what is APL (#580)
saleor/saleor-docs: docs/developer/extending/apps/architecture/apl.mdx

Saleor provides a simple persistence with App Metadata. Still, if App doesn't have a token and domain saved (which is typical for serverless architecture), it can't also call Saleor to reach Metadata. Hence need for simple, fast storage to save auth data.

Saleor doesn't require App to use any specific technology for persistence. That's why there are many implementations of APL available in the app-sdk package. And writing a custom one is easy.

# Communication diagram

This diagram explains flow how APL is used:

Following diagram describes how APL is used to store token received from Saleor during registration

sequenceDiagram
Saleor->>+App: Call /api/register with Auth Token and Domain header
App->>+APL: Save token and header
APL->>+External Database: Save in persistence layer

Follow diagram describes example of external webhook, that calls the App and should send data to Saleor

sequenceDiagram
External Service->>+App: Call endpoint with webhook and Saleor domain
App->>+APL: Get token for received domain
APL->>+External Database: Ask for token in persistence layer
External Database->>+APL: Provide saved token
APL->>+App: Provide saved token

> App->>+Saleor: Call graphQL query for related domain with retrieved token

Author: ![Zrzut ekranu 2022-10-24 o 17 08 21](https://user-images.githubusercontent.com/9268745/197560513-1282156b-3a00-4fa7-97ca-721139dfd76e.png)

Author: @peelar can we style mermaid rendered a little bit? Looks poor

Reviewer: I am afraid I don't have any experience with that, I would have to look into it to make sure it is consistent across the entire page. If you want anything specific to change, let me know. I will take a look some other time.

Author: I think its about using some defualt CSS/config on entire docs. I'd just use Saleor font and color palette to be consistent 

cc @mirekm ;)

Reviewer 2: Should be possible to do that: https://mermaid-js.github.io/mermaid/#/./theming

Dont make my mistake again and don't forget about the dark mode :D

Reviewer 2: After we complete the docs I could fight with the styles

Saleor provides a simple persistence with App Metadata. Still, if the App doesn't have a token and domain saved (which is typical for serverless architecture), it can't also call Saleor to reach Metadata. Hence need for simple, fast storage to save auth data.

Saleor doesn't require an App to use any specific technology for persistence. That's why there are many implementations of APL available in the `app-sdk` package. If they don't suffice, you can easily write a custom one.

# Communication diagram

Following diagram describes how APL is used to store token received from Saleor during registration

```mermaid
sequenceDiagram
Saleor->>+App: Call /api/register with Auth Token and Domain header
App->>+APL: Save token and header
APL->>+External Database: Save in persistence layer
```

The following diagram describes an example of an external webhook, that calls the App and sends the data to Saleor:

```mermaid
sequenceDiagram
External Service->>+App: Call endpoint with webhook and Saleor domain
App->>+APL: Get token for received domain
APL->>+External Database: Ask for token in persistence layer
External Database->>+APL: Provide saved token
APL->>+App: Provide saved token

> App->>+Saleor: Call graphQL query for related domain with retrieved token

Reviewer: ```suggestion
App->>+Saleor: Call GraphQL query for a related domain with retrieved token
```

---
title: APL - Auth Persistence Layer
---

APL is the core component of communication between the Saleor and the App.

During installation, Saleor sends AppToken, which the App uses to communicate with graphQL API. The App must save this token and target domain - but where?

# What is APL

APL is an abstraction on key/value persistence that holds domains (under which Saleor API is available) and tokens for communication.

Saleor provides a simple persistence with App Metadata. Still, if App doesn't have a token and domain saved (which is typical for serverless architecture), it can't also call Saleor to reach Metadata. Hence need for simple, fast storage to save auth data.

Saleor doesn't require App to use any specific technology for persistence. That's why there are many implementations of APL available in the app-sdk package. And writing a custom one is easy.

# Communication diagram

> This diagram explains flow how APL is used:

Reviewer: ```suggestion
This diagram explains the flow of how the APL is used:
```

> Following diagram describes how APL is used to store token received from Saleor during registration

Reviewer: ```suggestion
The following diagram describes how APL is used to store tokens received from Saleor during registration:
```

Is this redundant? Because we have:
> This diagram explains flow how APL is used:

just above.

Author: I think the latter should stay 👍

```mermaid
sequenceDiagram
Saleor->>+App: Call /api/register with Auth Token and Domain header
App->>+APL: Save token and header
APL->>+External Database: Save in persistence layer
```

> Follow diagram describes example of external webhook, that calls the App and should send data to Saleor

Reviewer: ```suggestion
The following diagram describes an example of an external webhook, that calls the App and sends the data to Saleor:
```

```mermaid
sequenceDiagram
External Service->>+App: Call endpoint with webhook and Saleor domain
App->>+APL: Get token for received domain
APL->>+External Database: Ask for token in persistence layer
External Database->>+APL: Provide saved token
APL->>+App: Provide saved token
App->>+Saleor: Call graphQL query for related domain with retrieved token
```

# Shape of the APL

```typescript
export interface AuthData {
  domain: string;
  token: string;
}

export type AplReadyResult =
  | {
      ready: true;
    }
  | {
      ready: false;
      error: Error;
    };

export type AplConfiguredResult =
  | {
      configured: true;
    }
  | {
      configured: false;
      error: Error;
    };

export interface APL {
  get: (domain: string) => Promise<AuthData | undefined>;
  set: (authData: AuthData) => Promise<void>;
  delete: (domain: string) => Promise<void>;
  getAll: () => Promise<AuthData[]>;
  /**
   * Inform that configuration is finished and correct
   */
  isReady: () => Promise<AplReadyResult>;
  isConfigured: () => Promise<AplConfiguredResult>;
}
```

> You can find the [APL shape in app-sdk](https://github.com/saleor/saleor-app-sdk/blob/main/src/APL/apl.ts)

Reviewer: ```suggestion
You can find the [APL shape in app-sdk](https://github.com/saleor/saleor-app-sdk/blob/main/src/APL/apl.ts).
```

# Single vs. Multi-tenant APLs

Saleor apps can work in both Single and Multi-tenant modes. Single-tenant apps can hardcode the Saleor domain and even a token, but they can perform only as an extension for one Saleor instance. That's how [App Template Gallery](https://docs.saleor.io/docs/3.x/dashboard/app-templates-gallery) works. If another Saleor instance is used, it requires a separate App deployment (with updated settings/env).

The preferred way is to write apps in a Multi-tenant manner. That means a single deployment can connect to multiple Saleor instances. To achieve that, App must ask some database for a token assigned to a specific Saleor domain.

# Available APL clients

There are several implementations of APL, including:

- local filesystem (Single tenant)
- Vercel (single tenant)
- Upstash (Multi-tenant)
- Firestore (Multi-tenant)

The updated [list of APLs is here](https://github.com/saleor/saleor-app-sdk/blob/main/docs/apl.md#available-apls)

# Writing custom APL

> Writing custom APL is easy - you only need to implement `APL` interface from app-sdk. An [example of a custom Redis APL is here](https://github.com/saleor/saleor-app-sdk/blob/main/docs/apl.md#example-implementation)

Reviewer: ```suggestion
Writing custom APL is easy - you only need to implement the `APL` interface from `app-sdk`. You can find an example of a custom Redis APL [here](https://github.com/saleor/saleor-app-sdk/blob/main/docs/apl.md#example-implementation).
```

# Developing an App without APL

> Technically, a simple App can be developed without APL. The token can be created via graphQL mutation or in Dashboard. The App can also decide to work only with a single Saleor instance - as a custom extension only for a single store.
> In this case, APL can be omitted, and App can store domain and tokens in env variables (or hardcode, which is not recommended).
> 
> However, some features in app-sdk are meant to be used together with APL interface. That's why it's recommended to follow this pattern.

Reviewer: ```suggestion
Technically, a simple App can be developed without APL. The token can be created via GraphQL mutation or in the Dashboard. The App can also decide to work only with a single Saleor instance - as a custom extension only for a single store.
In this case, APL can be omitted, and App can store domain and tokens in env variables (or hardcode, which is not recommended).

However, some features in `app-sdk` are meant to be used together with the APL interface. That's why it's recommended to follow this pattern.
```
